services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerContext.java

```diff
@@ -0,0 +1,170 @@
+package com.linkedin.venice.controller;
+
+import static com.linkedin.venice.controller.VeniceController.CONTROLLER_SERVICE_NAME;
+
+import com.linkedin.d2.balancer.D2Client;
+import com.linkedin.venice.acl.DynamicAccessController;
+import com.linkedin.venice.authorization.AuthorizerService;
+import com.linkedin.venice.client.store.ClientConfig;
+import com.linkedin.venice.controller.supersetschema.SupersetSchemaGenerator;
+import com.linkedin.venice.pubsub.api.PubSubClientsFactory;
+import com.linkedin.venice.service.ICProvider;
+import com.linkedin.venice.servicediscovery.ServiceDiscoveryAnnouncer;
+import com.linkedin.venice.stats.TehutiUtils;
+import com.linkedin.venice.utils.VeniceProperties;
+import io.tehuti.metrics.MetricsRepository;
+import java.util.Collections;
+import java.util.List;
+
+
+/**
+ * VeniceControllerContext contains dependencies required by {@link VeniceController}
+ */
+public class VeniceControllerContext {
```

Comment:
So the usual pattern is that the constructor args to a builder like this are the mandatory configs and everything else is optional.  It seems to be implied by this code that everything is optional, but I'm not sure that thats true?  My sense is that maybe most of these are optional (d2client, authorizerservice, etc.) but can a controller possibly work without the pubsubclientsfactory?

Reply:
A controller will not work without `PubSubClientsFactory`, as the consumer, producer and admin adapter factory are all from it. We use the default setting of `PubSubClientsFactory` with `ApacheKafkaAdapters`. In integration test, it comes from `PubSubBrokerFactory#getClientsFactory()` by
`System.getProperty(PUBSUB_BROKER_FACTORY_FQCN, "com.linkedin.venice.integration.utils.KafkaBrokerFactory");`, we plan to override this property if we want to use other `PubSubClientFactory`.

Reply:
Wait, we're using a jvm variable to configure this?  Is that in test only or production code?

Reply:
This is only for test only code. In production, this PR has changed the constructor of `VeniceController` and `VeniceServer` , specific `PubSubClientFactory` can be passed by our internal MP.

Reply:
Add null checking in `VeniceController` constructor to force passing instances in.